daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DaffodilUnparseContentHandler.scala

```diff
@@ -0,0 +1,309 @@
+  }
+
+  /**
+   * Uses Attributes, which is passed in to the startElement callback, to extract prefix mappings and
+   * populate the global prefixMapping
+   */
+  def mapPrefixMappingFromAttributesImpl(atts:Attributes): Unit = {
+    var i = 0
+    while (i < atts.getLength) {
+      val qName = atts.getQName(i)
+      if (qName.startsWith("xmlns")) {
+        val uri =  atts.getValue(i)
+        val prefix = if(qName.contains(":")) {
+          val pref = qName.split(":").last
+          pref
+        } else {
+         null // NamespaceBinding does not allow blanks so return null instead
+        }
+        prefixMapping = NamespaceBinding(prefix, uri, prefixMapping)
+      }
+      i += 1
+    }
+  }
+
+  override def startElement(uri: String, localName: String, qName: String, atts: Attributes): Unit = {
+    // we need to check if the characters data is all whitespace, if it is we drop the whitespace
+    // data, if it is not, it is an error as starting a new element with actual characterData means
+    // we haven't hit an endElement yet, which means we're in a complexElement and a complexElement
+    // cannot have character content
+    if (characterData.nonEmpty && !Misc.isAllWhitespace(characterData)) {
+      throw new IllegalContentWhereEventExpected("Non-whitespace characters in complex " +
+        "Element: " + characterData.toString
+      )
+    } else {
+      // reset since it was whitespace only
+      characterData.setLength(0)
+    }
+
+    if (!contentHandlerPrefixMappingUsed) {
+      // always pushes but doesn't always add a mapping since atts can be empty
+      prefixMappingTrackingStack.push(prefixMapping)
+      mapPrefixMappingFromAttributesImpl(atts)
+    }
+
+    if (!infosetEvent.isEmpty && infosetEvent.localName.isDefined) {
+      // we started another element while we were in the process of building a startElement
+      // this means the first element was complex and we are ready for the inputter queue
+      sendToInputter()
+    }
+    // use Attributes to determine xsi:nil value
+    val nilIn = atts.getIndex(XMLConstants.W3C_XML_SCHEMA_INSTANCE_NS_URI, "nil")
+    infosetEvent.nilValue = if (nilIn >= 0) {
+      val nilValue = atts.getValue(nilIn)
+      One(nilValue)
+    } else {
+      Nope
+    }
+    // set localName and namespaceURI
+    lazy val qNameArr = qName.split(":")
+    infosetEvent.localName =
+      if (localName.nonEmpty) {
+        One(localName)
+      } else if (qNameArr.length > 1) {
+        One(qNameArr.last)
+      } else if (qNameArr.nonEmpty) {
+        One(qNameArr.head)
+      } else {
+        Nope
+      }
+    infosetEvent.namespaceURI =
+      if (uri.nonEmpty) {
+        One(uri)
+      } else if (qNameArr.length > 1) { // has a prefix
+        // this use case is for the situation where the XMLReader doesn't pass in uri/localname,
+        // so prefix information must be determined via AttributeImpl attribute. This can usually
```

Comment:
Do you mean Attribute instead of AttributeImpl here? The parameter passed in in just Attribute.